tap_dynamodb/sync_strategies/log_based.py

```diff
@@ -83,6 +83,48 @@ def get_latest_seq_numbers(config, stream):
 
     return sequence_number_bookmarks
 
+
+def sync_shard(shard, seq_number_bookmarks, streams_client, stream_arn, projection, deserializer, table_name, stream_version, state):
+    seq_number = seq_number_bookmarks.get(shard['ShardId'])
+    if seq_number:
+        iterator_type = 'AFTER_SEQUENCE_NUMBER'
+    else:
+        iterator_type = 'TRIM_HORIZON'
+
+    rows_synced = 0
+
+    for record in get_shard_records(streams_client, stream_arn, shard, iterator_type, seq_number):
+        if record['eventName'] == 'REMOVE':
+            record_message = deserializer.deserialize_item(record['dynamodb']['Keys'])
+            record_message[SDC_DELETED_AT] = singer.utils.strftime(record['dynamodb']['ApproximateCreationDateTime'])
+        else:
+            record_message = deserializer.deserialize_item(record['dynamodb'].get('NewImage'))
+            if record_message is None:
+                LOGGER.fatal('Dynamo stream view type must be either "NEW_IMAGE" "NEW_AND_OLD_IMAGES"')
+                raise RuntimeError('Dynamo stream view type must be either "NEW_IMAGE" "NEW_AND_OLD_IMAGES"')
+            if projection is not None:
+                try:
+                    record_message = deserializer.apply_projection(record_message, projection)
+                except:
+                    LOGGER.fatal("Projection failed to apply: %s", projection)
+                    raise RuntimeError('Projection failed to apply: {}'.format(projection))
+
+        record_message = singer.RecordMessage(stream=table_name,
+                                            record=record_message,
+                                            version=stream_version)
+        singer.write_message(record_message)
+
+        rows_synced += 1
+
+        seq_number_bookmarks[shard['ShardId']] = record['dynamodb']['SequenceNumber']
+        state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
+
+        if rows_synced % WRITE_STATE_PERIOD == 0:
+            singer.write_state(state)
```

Comment:
I feel like `write_state` should be called at the end of each shard. Since the `rows_synced` isn't shared between shards anymore. Just to be sure that the state gets there once per shard.

Reply:
Good catch on this. Performing a `write_state` on every shard isn't the best because there are often a dozen shards that are completely empty causing us to send out the same state message 12 times without anything changing. I've moved this `if rows_synced % WRITE_STATE_PERIOD == 0:` outside of `sync_shard()` and it should perform the same as it did previously.

Reply:
We chatted. Turns out that shards are like the Inception Totems and we need to see one we know so that the world makes sense again.